File: unitree_slam_example_new/global_nav/plan_nav_dynamic_replan.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <limits>
#include <queue>
#include <regex>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

struct Config {
  std::string map_path;
  std::string obstacles_json;
  std::string output_path;
  double start_x = 0.0;
  double start_y = 0.0;
  double goal_x = 0.0;
  double goal_y = 0.0;
  double goal_yaw = 0.0;
  double max_segment = 10.0;
  double dyn_margin = 0.10;
  double cooldown = 1.0;
  int confirm_frames = 3;
};

struct Map2D {
  int width = 0;
  int height = 0;
  double origin_x = 0.0;
  double origin_y = 0.0;
  double resolution = 0.1;
  std::vector<uint8_t> occ;
};

struct Node {
  int x;
  int y;
};

struct DynObstacle {
  double x;
  double y;
  double r;
};

struct DynFrame {
  double t = 0.0;
  std::vector<DynObstacle> obstacles;
};

struct ReplanEvent {
  double t = 0.0;
  int trigger_idx = 0;
  int remaining_waypoints_after_replan = 0;
  int num_dyn_obstacles = 0;
};

static void PrintUsage() {
  std::cout << "Usage:\n"
            << "  plan_nav_dynamic_replan_cpp --map <map.txt> --start <x> <y> --goal <x> <y> "
               "--obstacles-json <obs.json> --output <path.json> "
               "[--goal-yaw 0.0] [--max-segment 10.0] [--dyn-margin 0.10] [--cooldown 1.0] [--confirm-frames 3]\n";
}

static bool ParseArgs(int argc, char** argv, Config& cfg) {
  for (int i = 1; i < argc; ++i) {
    std::string a = argv[i];
    auto need = [&](const std::string& n) {
      if (i + 1 >= argc) throw std::runtime_error("Missing value for " + n);
      return std::string(argv[++i]);
    };

    if (a == "--map") cfg.map_path = need(a);
    else if (a == "--obstacles-json") cfg.obstacles_json = need(a);
    else if (a == "--output") cfg.output_path = need(a);
    else if (a == "--start") {
      cfg.start_x = std::stod(need(a));
      cfg.start_y = std::stod(need(a));
    } else if (a == "--goal") {
      cfg.goal_x = std::stod(need(a));
      cfg.goal_y = std::stod(need(a));
    } else if (a == "--goal-yaw") cfg.goal_yaw = std::stod(need(a));
    else if (a == "--max-segment") cfg.max_segment = std::stod(need(a));
    else if (a == "--dyn-margin") cfg.dyn_margin = std::stod(need(a));
    else if (a == "--cooldown") cfg.cooldown = std::stod(need(a));
    else if (a == "--confirm-frames") cfg.confirm_frames = std::stoi(need(a));
    else {
      std::cerr << "Unknown arg: " << a << "\n";
      return false;
    }
  }

  return !cfg.map_path.empty() && !cfg.output_path.empty() && !cfg.obstacles_json.empty();
}

static size_t Key(int x, int y) { return (static_cast<size_t>(y) << 32) ^ static_cast<uint32_t>(x); }

static bool InBounds(const Map2D& m, int x, int y) {
  return x >= 0 && x < m.width && y >= 0 && y < m.height;
}

static bool IsFree(const Map2D& m, int x, int y) {
  return InBounds(m, x, y) && m.occ[(size_t)y * m.width + x] == 0;
}

static Node WorldToGrid(const Map2D& m, double x, double y) {
  int gx = static_cast<int>(std::floor((x - m.origin_x) / m.resolution));
  int gy = static_cast<int>(std::floor((y - m.origin_y) / m.resolution));
  return {gx, gy};
}

static std::pair<double, double> GridToWorld(const Map2D& m, int gx, int gy) {
  double x = m.origin_x + (gx + 0.5) * m.resolution;
  double y = m.origin_y + (gy + 0.5) * m.resolution;
  return {x, y};
}

static bool LoadMap(const std::string& path, Map2D& m) {
  std::ifstream ifs(path);
  if (!ifs) return false;

  ifs >> m.width >> m.height;
  ifs >> m.origin_x >> m.origin_y >> m.resolution;
  std::string row;
  std::getline(ifs, row);

  m.occ.assign((size_t)m.width * m.height, 0);
  for (int y = 0; y < m.height; ++y) {
    if (!std::getline(ifs, row)) return false;
    if ((int)row.size() < m.width) return false;
    for (int x = 0; x < m.width; ++x) {
      m.occ[(size_t)y * m.width + x] = (row[x] == '1') ? 1 : 0;
    }
  }
  return true;
}

static bool LineIsFree(const Map2D& m, Node a, Node b) {
  int x0 = a.x, y0 = a.y;
  int x1 = b.x, y1 = b.y;
  int dx = std::abs(x1 - x0), dy = std::abs(y1 - y0);
  int sx = (x0 < x1) ? 1 : -1;
  int sy = (y0 < y1) ? 1 : -1;
  int err = dx - dy;

  int x = x0, y = y0;
  while (true) {
    if (!IsFree(m, x, y)) return false;
    if (x == x1 && y == y1) break;
    int e2 = 2 * err;
    if (e2 > -dy) {
      err -= dy;
      x += sx;
    }
    if (e2 < dx) {
      err += dx;
      y += sy;
    }
  }
  return true;
}

static std::vector<Node> AStar(const Map2D& m, Node start, Node goal) {
  if (!IsFree(m, start.x, start.y)) throw std::runtime_error("Start is occupied/out-of-bounds");
  if (!IsFree(m, goal.x, goal.y)) throw std::runtime_error("Goal is occupied/out-of-bounds");

  struct QN {
    double f;
    Node n;
    bool operator<(const QN& o) const { return f > o.f; }
  };

  auto h = [](Node a, Node b) { return std::hypot((double)a.x - b.x, (double)a.y - b.y); };
  const int dxs[8] = {-1, 1, 0, 0, -1, -1, 1, 1};
  const int dys[8] = {0, 0, -1, 1, -1, 1, -1, 1};
  const double cst[8] = {1.0, 1.0, 1.0, 1.0, std::sqrt(2.0), std::sqrt(2.0), std::sqrt(2.0), std::sqrt(2.0)};

  std::priority_queue<QN> pq;
  std::unordered_map<size_t, double> g;
  std::unordered_map<size_t, Node> parent;
  std::unordered_map<size_t, bool> closed;

  g[Key(start.x, start.y)] = 0.0;
  pq.push({h(start, goal), start});

  while (!pq.empty()) {
    Node cur = pq.top().n;
    pq.pop();
    size_t ck = Key(cur.x, cur.y);
    if (closed[ck]) continue;
    closed[ck] = true;

    if (cur.x == goal.x && cur.y == goal.y) {
      std::vector<Node> path;
      Node t = goal;
      path.push_back(t);
      while (!(t.x == start.x && t.y == start.y)) {
        t = parent.at(Key(t.x, t.y));
        path.push_back(t);
      }
      std::reverse(path.begin(), path.end());
      return path;
    }

    for (int k = 0; k < 8; ++k) {
      int nx = cur.x + dxs[k], ny = cur.y + dys[k];
      Node nxt{nx, ny};
      if (!IsFree(m, nx, ny)) continue;
      if (dxs[k] != 0 && dys[k] != 0) {
        if (!IsFree(m, cur.x + dxs[k], cur.y) || !IsFree(m, cur.x, cur.y + dys[k])) continue;
      }

      size_t nk = Key(nx, ny);
      double ng = g[ck] + cst[k];
      auto it = g.find(nk);
      if (it == g.end() || ng < it->second) {
        g[nk] = ng;
        parent[nk] = cur;
        pq.push({ng + h(nxt, goal), nxt});
      }
    }
  }

  throw std::runtime_error("No path found");
}

static std::vector<Node> Sparsify(const Map2D& m, const std::vector<Node>& path, double max_seg_m) {
  if (path.size() <= 2) return path;
  const double max_cells = max_seg_m / m.resolution;
  std::vector<Node> out;
  out.push_back(path.front());

  size_t i = 0;
  while (i + 1 < path.size()) {
    size_t best = i + 1;
    for (size_t j = i + 1; j < path.size(); ++j) {
      double dx = (double)path[j].x - path[i].x;
      double dy = (double)path[j].y - path[i].y;
      if (std::hypot(dx, dy) > max_cells) break;
      if (LineIsFree(m, path[i], path[j])) best = j;
      else break;
    }
    out.push_back(path[best]);
    i = best;
  }
  return out;
}

static void InflateDynamic(Map2D& m, const std::vector<DynObstacle>& obs, double extra_margin) {
  auto occ0 = m.occ;
  auto idx = [&m](int x, int y) { return (size_t)y * m.width + x; };

  for (const auto& o : obs) {
    const int cells = (int)std::ceil((o.r + extra_margin) / m.resolution);
    Node c = WorldToGrid(m, o.x, o.y);

    int y0 = std::max(0, c.y - cells), y1 = std::min(m.height - 1, c.y + cells);
    int x0 = std::max(0, c.x - cells), x1 = std::min(m.width - 1, c.x + cells);

    for (int y = y0; y <= y1; ++y) {
      for (int x = x0; x <= x1; ++x) {
        int dx = x - c.x, dy = y - c.y;
        if (dx * dx + dy * dy <= cells * cells) m.occ[idx(x, y)] = 1;
      }
    }
  }
}

static bool RemainingBlocked(const Map2D& m, const std::vector<Node>& sparse, int from_idx) {
  if (from_idx >= (int)sparse.size() - 1) return false;
  for (int i = from_idx; i < (int)sparse.size() - 1; ++i) {
    if (!LineIsFree(m, sparse[i], sparse[i + 1])) return true;
  }
  return false;
}

static int NearestIndex(const std::vector<Node>& path, const Node& cur) {
  int best = 0;
  double best_d = std::numeric_limits<double>::infinity();
  for (int i = 0; i < (int)path.size(); ++i) {
    double dx = (double)path[i].x - cur.x;
    double dy = (double)path[i].y - cur.y;
    double d = dx * dx + dy * dy;
    if (d < best_d) {
      best_d = d;
      best = i;
    }
  }
  return best;
}

// Minimal JSON parser for expected schema only.
static std::vector<DynFrame> LoadFrames(const std::string& path) {
  std::ifstream ifs(path);
  if (!ifs) throw std::runtime_error("Cannot open obstacles json: " + path);
  std::stringstream buf;
  buf << ifs.rdbuf();
  std::string s = buf.str();

  std::vector<DynFrame> frames;

  std::regex frame_re(R"(\{\s*"t"\s*:\s*([-+0-9.eE]+)\s*,\s*"obstacles"\s*:\s*\[(.*?)\]\s*\})");
  std::regex obs_re(R"(\{\s*"x"\s*:\s*([-+0-9.eE]+)\s*,\s*"y"\s*:\s*([-+0-9.eE]+)(?:\s*,\s*"r"\s*:\s*([-+0-9.eE]+))?\s*\})");

  auto fb = std::sregex_iterator(s.begin(), s.end(), frame_re);
  auto fe = std::sregex_iterator();
  for (auto it = fb; it != fe; ++it) {
    DynFrame f;
    f.t = std::stod((*it)[1].str());
    std::string obs_block = (*it)[2].str();

    auto ob = std::sregex_iterator(obs_block.begin(), obs_block.end(), obs_re);
    auto oe = std::sregex_iterator();
    for (auto oit = ob; oit != oe; ++oit) {
      DynObstacle o;
      o.x = std::stod((*oit)[1].str());
      o.y = std::stod((*oit)[2].str());
      o.r = (*oit)[3].matched ? std::stod((*oit)[3].str()) : 0.3;
      f.obstacles.push_back(o);
    }
    frames.push_back(f);
  }

  std::sort(frames.begin(), frames.end(), [](const DynFrame& a, const DynFrame& b) { return a.t < b.t; });
  return frames;
}

int main(int argc, char** argv) {
  Config cfg;
  try {
    if (!ParseArgs(argc, argv, cfg)) {
      PrintUsage();
      return 1;
    }

    Map2D static_map;
    if (!LoadMap(cfg.map_path, static_map)) {
      throw std::runtime_error("Failed to load map text file: " + cfg.map_path);
    }

    Node start = WorldToGrid(static_map, cfg.start_x, cfg.start_y);
    Node goal = WorldToGrid(static_map, cfg.goal_x, cfg.goal_y);

    auto dense = AStar(static_map, start, goal);
    auto sparse = Sparsify(static_map, dense, cfg.max_segment);

    auto frames = LoadFrames(cfg.obstacles_json);

    int blocked_count = 0;
    double last_replan_t = -1e9;
    std::vector<ReplanEvent> events;
    int cur_idx = 0;

    for (const auto& f : frames) {
      Map2D dyn = static_map;
      InflateDynamic(dyn, f.obstacles, cfg.dyn_margin);

      if (sparse.empty()) break;
      cur_idx = std::min(cur_idx, (int)sparse.size() - 1);
      cur_idx = NearestIndex(sparse, sparse[cur_idx]);

      bool blocked = RemainingBlocked(dyn, sparse, cur_idx);
      blocked_count = blocked ? (blocked_count + 1) : 0;

      bool should_replan =
          blocked_count >= cfg.confirm_frames &&
          (f.t - last_replan_t) >= cfg.cooldown &&
          cur_idx < (int)sparse.size() - 1;

      if (should_replan) {
        Node repl_start = sparse[cur_idx];
        auto d2 = AStar(dyn, repl_start, goal);
        auto s2 = Sparsify(dyn, d2, cfg.max_segment);

        std::vector<Node> merged;
        for (int i = 0; i < cur_idx; ++i) merged.push_back(sparse[i]);
        if (merged.empty() || (merged.back().x != s2.front().x || merged.back().y != s2.front().y)) {
          merged.insert(merged.end(), s2.begin(), s2.end());
        } else {
          merged.insert(merged.end(), s2.begin() + 1, s2.end());
        }
        sparse.swap(merged);

        events.push_back({f.t, cur_idx, (int)sparse.size() - cur_idx, (int)f.obstacles.size()});
        last_replan_t = f.t;
        blocked_count = 0;
      }

      if (cur_idx < (int)sparse.size() - 1 && LineIsFree(dyn, sparse[cur_idx], sparse[cur_idx + 1])) {
        cur_idx++;
      }
    }

    std::filesystem::create_directories(std::filesystem::path(cfg.output_path).parent_path());
    std::ofstream ofs(cfg.output_path);
    if (!ofs) throw std::runtime_error("Cannot write output json");

    ofs << "{\n";
    ofs << "  \"map\": \"" << cfg.map_path << "\",\n";
    ofs << "  \"start\": {\"x\": " << cfg.start_x << ", \"y\": " << cfg.start_y << "},\n";
    ofs << "  \"goal\": {\"x\": " << cfg.goal_x << ", \"y\": " << cfg.goal_y
        << ", \"yaw\": " << cfg.goal_yaw << "},\n";
    ofs << "  \"max_segment\": " << cfg.max_segment << ",\n";
    ofs << "  \"num_waypoints\": " << sparse.size() << ",\n";
    ofs << "  \"num_replans\": " << events.size() << ",\n";
    ofs << "  \"replan_events\": [\n";
    for (size_t i = 0; i < events.size(); ++i) {
      const auto& e = events[i];
      ofs << "    {\"t\": " << e.t
          << ", \"trigger_idx\": " << e.trigger_idx
          << ", \"remaining_waypoints_after_replan\": " << e.remaining_waypoints_after_replan
          << ", \"num_dyn_obstacles\": " << e.num_dyn_obstacles << "}";
      if (i + 1 < events.size()) ofs << ",";
      ofs << "\n";
    }
    ofs << "  ],\n";

    ofs << "  \"waypoints\": [\n";
    for (size_t i = 0; i < sparse.size(); ++i) {
      auto [wx, wy] = GridToWorld(static_map, sparse[i].x, sparse[i].y);
      double qx = 0.0, qy = 0.0, qz = 0.0, qw = 1.0;
      if (i + 1 < sparse.size()) {
        auto [nx, ny] = GridToWorld(static_map, sparse[i + 1].x, sparse[i + 1].y);
        double yaw = std::atan2(ny - wy, nx - wx);
        qz = std::sin(yaw * 0.5);
        qw = std::cos(yaw * 0.5);
      } else {
        qz = std::sin(cfg.goal_yaw * 0.5);
        qw = std::cos(cfg.goal_yaw * 0.5);
      }

      ofs << "    {\"index\": " << i
          << ", \"x\": " << wx
          << ", \"y\": " << wy
          << ", \"z\": 0.0"
          << ", \"qx\": " << qx
          << ", \"qy\": " << qy
          << ", \"qz\": " << qz
          << ", \"qw\": " << qw << "}";
      if (i + 1 < sparse.size()) ofs << ",";
      ofs << "\n";
    }
    ofs << "  ]\n";
    ofs << "}\n";

    std::cout << "Dynamic path saved: " << cfg.output_path << "\n";
    std::cout << "Final waypoints: " << sparse.size() << ", replans: " << events.size() << "\n";

  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << "\n";
    return 1;
  }

  return 0;
}

File: unitree_slam_example_new/global_nav/plan_nav_dynamic_replan.py
#!/usr/bin/env python3
"""
Event-triggered dynamic replanning on top of static global planning.

Core idea:
- Start from a static global sparse path.
- At each time step, inject dynamic obstacles into a temporary occupancy map.
- If remaining path is blocked by new obstacles, trigger replanning from current robot position.
- Add anti-jitter protections: block confirmation + replan cooldown.

Input obstacle JSON example:
{
  "frames": [
    {"t": 0.0, "obstacles": [{"x": 0.5, "y": 1.2, "r": 0.4}]},
    {"t": 0.5, "obstacles": [{"x": 0.6, "y": 1.2, "r": 0.4}]}
  ]
}
"""

from __future__ import annotations

import argparse
import heapq
import json
import math
from dataclasses import dataclass
from pathlib import Path

import numpy as np


@dataclass
class DynObstacle:
    x: float
    y: float
    r: float


def world_to_grid(x: float, y: float, origin: np.ndarray, res: float) -> tuple[int, int]:
    gx = int(math.floor((x - origin[0]) / res))
    gy = int(math.floor((y - origin[1]) / res))
    return gx, gy


def grid_to_world(gx: int, gy: int, origin: np.ndarray, res: float) -> tuple[float, float]:
    x = origin[0] + (gx + 0.5) * res
    y = origin[1] + (gy + 0.5) * res
    return x, y


def in_bounds(gx: int, gy: int, w: int, h: int) -> bool:
    return 0 <= gx < w and 0 <= gy < h


def line_is_free(occ: np.ndarray, a: tuple[int, int], b: tuple[int, int]) -> bool:
    x0, y0 = a
    x1, y1 = b
    dx = abs(x1 - x0)
    dy = abs(y1 - y0)
    sx = 1 if x0 < x1 else -1
    sy = 1 if y0 < y1 else -1
    err = dx - dy

    h, w = occ.shape
    x, y = x0, y0
    while True:
        if not in_bounds(x, y, w, h) or occ[y, x] != 0:
            return False
        if x == x1 and y == y1:
            break
        e2 = 2 * err
        if e2 > -dy:
            err -= dy
            x += sx
        if e2 < dx:
            err += dx
            y += sy
    return True


def astar(occ: np.ndarray, start: tuple[int, int], goal: tuple[int, int]) -> list[tuple[int, int]]:
    h, w = occ.shape
    if not in_bounds(start[0], start[1], w, h) or not in_bounds(goal[0], goal[1], w, h):
        raise ValueError("Start or goal out of map bounds")
    if occ[start[1], start[0]] != 0:
        raise ValueError("Start in obstacle")
    if occ[goal[1], goal[0]] != 0:
        raise ValueError("Goal in obstacle")

    neigh = [
        (-1, 0, 1.0), (1, 0, 1.0), (0, -1, 1.0), (0, 1, 1.0),
        (-1, -1, math.sqrt(2.0)), (-1, 1, math.sqrt(2.0)),
        (1, -1, math.sqrt(2.0)), (1, 1, math.sqrt(2.0)),
    ]

    def heuristic(a: tuple[int, int], b: tuple[int, int]) -> float:
        return math.hypot(a[0] - b[0], a[1] - b[1])

    g_cost: dict[tuple[int, int], float] = {start: 0.0}
    parent: dict[tuple[int, int], tuple[int, int]] = {}
    open_heap: list[tuple[float, tuple[int, int]]] = [(heuristic(start, goal), start)]
    closed: set[tuple[int, int]] = set()

    while open_heap:
        _, cur = heapq.heappop(open_heap)
        if cur in closed:
            continue
        closed.add(cur)

        if cur == goal:
            path = [cur]
            while cur in parent:
                cur = parent[cur]
                path.append(cur)
            path.reverse()
            return path

        cx, cy = cur
        for dx, dy, step in neigh:
            nx, ny = cx + dx, cy + dy
            nxt = (nx, ny)
            if not in_bounds(nx, ny, w, h) or occ[ny, nx] != 0:
                continue
            if dx != 0 and dy != 0 and (occ[cy, nx] != 0 or occ[ny, cx] != 0):
                continue

            ng = g_cost[cur] + step
            if ng < g_cost.get(nxt, float("inf")):
                g_cost[nxt] = ng
                parent[nxt] = cur
                heapq.heappush(open_heap, (ng + heuristic(nxt, goal), nxt))

    raise RuntimeError("No path found")


def sparsify_path(occ: np.ndarray, path: list[tuple[int, int]], res: float, max_segment_m: float) -> list[tuple[int, int]]:
    if len(path) <= 2:
        return path
    max_cells = int(math.floor(max_segment_m / res))
    out = [path[0]]
    i = 0
    while i < len(path) - 1:
        best = i + 1
        for j in range(i + 1, len(path)):
            if math.hypot(path[j][0] - path[i][0], path[j][1] - path[i][1]) > max_cells:
                break
            if line_is_free(occ, path[i], path[j]):
                best = j
            else:
                break
        out.append(path[best])
        i = best
    return out


def yaw_quat_from_to(a_xy: tuple[float, float], b_xy: tuple[float, float]) -> tuple[float, float, float, float]:
    yaw = math.atan2(b_xy[1] - a_xy[1], b_xy[0] - a_xy[0])
    return 0.0, 0.0, math.sin(yaw * 0.5), math.cos(yaw * 0.5)


def inflate_dynamic_obstacles(
    static_occ: np.ndarray,
    origin: np.ndarray,
    res: float,
    obstacles: list[DynObstacle],
    extra_margin: float,
) -> np.ndarray:
    occ = static_occ.copy()
    h, w = occ.shape
    for obs in obstacles:
        cells = int(math.ceil((obs.r + extra_margin) / res))
        cx, cy = world_to_grid(obs.x, obs.y, origin, res)
        y0, y1 = max(0, cy - cells), min(h - 1, cy + cells)
        x0, x1 = max(0, cx - cells), min(w - 1, cx + cells)
        for yy in range(y0, y1 + 1):
            for xx in range(x0, x1 + 1):
                if (xx - cx) ** 2 + (yy - cy) ** 2 <= cells * cells:
                    occ[yy, xx] = 1
    return occ


def remaining_path_blocked(occ: np.ndarray, path: list[tuple[int, int]], from_idx: int) -> bool:
    if from_idx >= len(path) - 1:
        return False
    for i in range(from_idx, len(path) - 1):
        if not line_is_free(occ, path[i], path[i + 1]):
            return True
    return False


def nearest_path_index(path: list[tuple[int, int]], cur: tuple[int, int]) -> int:
    best_i = 0
    best_d = float("inf")
    for i, p in enumerate(path):
        d = (p[0] - cur[0]) ** 2 + (p[1] - cur[1]) ** 2
        if d < best_d:
            best_d = d
            best_i = i
    return best_i


def compose_waypoint_json(
    map_path: str,
    start_xy: tuple[float, float],
    goal_xy: tuple[float, float],
    goal_yaw: float,
    max_segment: float,
    sparse_grid: list[tuple[int, int]],
    origin: np.ndarray,
    res: float,
    replan_events: list[dict],
) -> dict:
    world_pts = [grid_to_world(px, py, origin, res) for px, py in sparse_grid]
    waypoints = []
    for i, (x, y) in enumerate(world_pts):
        if i < len(world_pts) - 1:
            qx, qy, qz, qw = yaw_quat_from_to((x, y), world_pts[i + 1])
        else:
            qx, qy, qz, qw = 0.0, 0.0, math.sin(goal_yaw * 0.5), math.cos(goal_yaw * 0.5)
        waypoints.append({
            "index": i,
            "x": x,
            "y": y,
            "z": 0.0,
            "qx": qx,
            "qy": qy,
            "qz": qz,
            "qw": qw,
        })

    return {
        "map": str(Path(map_path).resolve()),
        "start": {"x": start_xy[0], "y": start_xy[1]},
        "goal": {"x": goal_xy[0], "y": goal_xy[1], "yaw": goal_yaw},
        "max_segment": max_segment,
        "num_waypoints": len(waypoints),
        "num_replans": len(replan_events),
        "replan_events": replan_events,
        "waypoints": waypoints,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Dynamic event-triggered replan")
    parser.add_argument("--map", required=True, help="Input .npz map")
    parser.add_argument("--start", nargs=2, type=float, required=True, metavar=("X", "Y"))
    parser.add_argument("--goal", nargs=2, type=float, required=True, metavar=("X", "Y"))
    parser.add_argument("--goal-yaw", type=float, default=0.0)
    parser.add_argument("--max-segment", type=float, default=10.0)
    parser.add_argument("--obstacles-json", required=True, help="Dynamic obstacles frames JSON")
    parser.add_argument("--dyn-margin", type=float, default=0.10, help="Extra dynamic obstacle safety margin (m)")
    parser.add_argument("--cooldown", type=float, default=1.0, help="Min time between replans (s)")
    parser.add_argument("--confirm-frames", type=int, default=3, help="Consecutive blocked frames before replan")
    parser.add_argument("--output", required=True, help="Output JSON")
    args = parser.parse_args()

    data = np.load(args.map)
    static_occ = data["occupancy"].astype(np.uint8)
    origin = data["origin"].astype(np.float64)
    res = float(data["resolution"][0])

    start = world_to_grid(args.start[0], args.start[1], origin, res)
    goal = world_to_grid(args.goal[0], args.goal[1], origin, res)

    dense = astar(static_occ, start, goal)
    sparse = sparsify_path(static_occ, dense, res, args.max_segment)

    obs_obj = json.loads(Path(args.obstacles_json).read_text(encoding="utf-8"))
    frames = obs_obj.get("frames", [])

    blocked_count = 0
    last_replan_t = -1e9
    replan_events: list[dict] = []

    # Simulated progression: robot follows sparse path index over time.
    cur_idx = 0

    for frame in frames:
        t = float(frame.get("t", 0.0))
        obstacles = [DynObstacle(float(o["x"]), float(o["y"]), float(o.get("r", 0.3))) for o in frame.get("obstacles", [])]

        dyn_occ = inflate_dynamic_obstacles(static_occ, origin, res, obstacles, args.dyn_margin)

        # Update nearest progress index to avoid replanning from stale index.
        cur_idx = min(cur_idx, len(sparse) - 1)
        cur_idx = nearest_path_index(sparse, sparse[cur_idx])

        blocked = remaining_path_blocked(dyn_occ, sparse, cur_idx)
        blocked_count = blocked_count + 1 if blocked else 0

        should_replan = (
            blocked_count >= args.confirm_frames
            and (t - last_replan_t) >= args.cooldown
            and cur_idx < len(sparse) - 1
        )

        if should_replan:
            repl_start = sparse[cur_idx]
            dense_new = astar(dyn_occ, repl_start, goal)
            sparse_new = sparsify_path(dyn_occ, dense_new, res, args.max_segment)

            # splice: keep traveled prefix, replace remaining path
            prefix = sparse[:cur_idx]
            if not prefix or prefix[-1] != sparse_new[0]:
                sparse = prefix + sparse_new
            else:
                sparse = prefix + sparse_new[1:]

            replan_events.append({
                "t": t,
                "trigger_idx": cur_idx,
                "remaining_waypoints_after_replan": len(sparse) - cur_idx,
                "num_dyn_obstacles": len(obstacles),
            })
            last_replan_t = t
            blocked_count = 0

        # simple simulated movement: advance one waypoint if line is free
        if cur_idx < len(sparse) - 1 and line_is_free(dyn_occ, sparse[cur_idx], sparse[cur_idx + 1]):
            cur_idx += 1

    result = compose_waypoint_json(
        map_path=args.map,
        start_xy=(args.start[0], args.start[1]),
        goal_xy=(args.goal[0], args.goal[1]),
        goal_yaw=args.goal_yaw,
        max_segment=args.max_segment,
        sparse_grid=sparse,
        origin=origin,
        res=res,
        replan_events=replan_events,
    )

    out = Path(args.output)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(json.dumps(result, ensure_ascii=False, indent=2), encoding="utf-8")

    print("Dynamic planning finished:", out)
    print("Final waypoints:", result["num_waypoints"], "Replans:", result["num_replans"])


if __name__ == "__main__":
    main()
